src/wallet/wallet.cpp

```diff
@@ -82,15 +82,63 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
     return nullptr;
 }
 
-// Custom deleter for shared_ptr<CWallet>.
-static void ReleaseWallet(CWallet* wallet)
+static Mutex g_wallet_release_mutex;
+static std::condition_variable g_wallet_release_cv;
+static std::set<CWallet*> g_unloading_wallet_set;
+
+// Unregister and delete the wallet right after BlockUntilSyncedToCurrentChain
+// so that it's in sync with the current chainstate.
+static void DeleteWallet(CWallet* wallet)
 {
     wallet->WalletLogPrintf("Releasing wallet\n");
     wallet->BlockUntilSyncedToCurrentChain();
     wallet->Flush();
+    UnregisterValidationInterface(wallet);
     delete wallet;
 }
 
+// Custom deleter for shared_ptr<CWallet>.
+static void ReleaseWallet(CWallet* wallet)
+{
+    bool unloading{false};
```

Comment:
Is there a reason this needs to delegate deletion? It seems if you removed the DeleteWallet call from UnloadWallet, you could simplify this function to just:

```c++
DeleteWallet(wallet);
LOCK(g_wallet_release_mutex);
g_unloading_wallet_set.erase(wallet);
g_wallet_release_cv.notify_all();
```

Reply:
No strong reason, just thought the unload burden/time should be left to the `unloadwallet` caller. The way you suggest both the `unloadwallet` caller and the last wallet client wait for the complete unload.

Reply:
> just thought the unload burden/time should be left to the unloadwallet caller. 

Interesting. That seems like a legitimate optimization for the last RPC using the wallet. Personally, I think I'd still choose the simplification over the optimization, but if you want to keep the optimization it'd be good to have a comment saying why it's better to delete on the unload thread than the release thread.